pkg/shed/index.go

```diff
@@ -322,102 +341,66 @@ func (f Index) Iterate(fn IndexIterFunc, options *IterateOptions) (err error) {
+// If Index has no elements, a ErrNotFound error is returned.
 func (f Index) First(prefix []byte) (i Item, err error) {
-	it := f.db.NewIterator()
-	defer it.Release()
-
 	totalPrefix := append(f.prefix, prefix...)
-	it.Seek(totalPrefix)
-
-	return f.itemFromIterator(it, totalPrefix)
+	k, v, err := f.db.First(totalPrefix)
+	if err != nil {
+		return i, err
+	}
+	return f.itemFromKeyValue(k, v, totalPrefix)
 }
 
-// itemFromIterator returns the Item from the current iterator position.
+// itemFromKeyValue returns the Item from the current iterator position.
 // If the complete encoded key does not start with totalPrefix,
-// leveldb.ErrNotFound is returned. Value for totalPrefix must start with
+// ErrNotFound is returned. Value for totalPrefix must start with
 // Index prefix.
-func (f Index) itemFromIterator(it iterator.Iterator, totalPrefix []byte) (i Item, err error) {
-	key := it.Key()
+func (f Index) itemFromKeyValue(key []byte, value []byte, totalPrefix []byte) (i Item, err error) {
 	if !bytes.HasPrefix(key, totalPrefix) {
-		return i, leveldb.ErrNotFound
+		return i, badger.ErrKeyNotFound
 	}
-	// create a copy of key byte slice not to share leveldb underlaying slice array
+	// create a copy of key byte slice not to share badger underlaying slice array
 	keyItem, err := f.decodeKeyFunc(append([]byte(nil), key...))
 	if err != nil {
 		f.logger.Debugf("error decoding key in itemFromIterator. Error: %s", err.Error())
 		return i, err
 	}
-	// create a copy of value byte slice not to share leveldb underlaying slice array
-	valueItem, err := f.decodeValueFunc(keyItem, append([]byte(nil), it.Value()...))
+	// create a copy of value byte slice not to share badger underlaying slice array
+	valueItem, err := f.decodeValueFunc(keyItem, append([]byte(nil), value...))
 	if err != nil {
 		f.logger.Debugf("error decoding value in itemFromIterator. Error: %s", err.Error())
 		return i, err
 	}
-	return keyItem.Merge(valueItem), it.Error()
+	return keyItem.Merge(valueItem), nil
 }
 
 // Last returns the last item in the Index which encoded key starts with a prefix.
 // If the prefix is nil, the last element of the whole index is returned.
-// If Index has no elements, a leveldb.ErrNotFound error is returned.
+// If Index has no elements, a ErrNotFound error is returned.
 func (f Index) Last(prefix []byte) (i Item, err error) {
-	it := f.db.NewIterator()
-	defer it.Release()
-
-	// get the next prefix in line
-	// since leveldb iterator Seek seeks to the
-	// next key if the key that it seeks to is not found
-	// and by getting the previous key, the last one for the
-	// actual prefix is found
-	nextPrefix := incByteSlice(prefix)
-	l := len(prefix)
-
-	if l > 0 && nextPrefix != nil {
-		it.Seek(append(f.prefix, nextPrefix...))
-		it.Prev()
-	} else {
-		it.Last()
-	}
-
 	totalPrefix := append(f.prefix, prefix...)
-	return f.itemFromIterator(it, totalPrefix)
+	k, v, err := f.db.Last(totalPrefix)
```

Comment:
This change replaces an efficient implementation with one very inefficient. Related to comment to `db.Last`.

Reply:
As i said in the older commit.. there is no way to do a Previous() while doing a iteration. Let me see how else can i fix this.

Reply:
Couldn't the `IteratorOptions.Reverse` option be used to get the Prev functionality?

Reply:
fixed this.